### DISCUSSION TITLE
Measuring post phase shift on digital basis

### DISCUSSION THREAD

**Discussion by juanboschero** [pasqal-io/Pulser#505] [Q&A] [answered]
Hi all,

I am quite new to the world of quantum computing with Rydberg atoms and I am trying to use Pulser to pulse elaborate qubit gates. I am trying to incorporate the single qubit Z rotation gate in the digital basis using a virtual post phase shift  (as described by the docs) but upon simulating the result in the digital basis, I see that nothing changes. Is there something that I am fundamentally missing or is there another way to incorporate a Z gate for it to appear in the simulation?

Thank you very much!

**Comment by HGSilveri** [comment]
Hi @Santopelufincho, welcome ! 
The virtual-Z gate implementation relies on the fact that Z-rotations right before a measurement have no influence on the outcomes, so in fact this last rotation is not even done. Could it be that the last thing you are doing is the Z-rotation? 
Feel free to share a snippet of your code so that we can see what you are trying out and give more concrete tips.

**Reply by juanboschero** [reply]
Hi @HGSilveri, 

Thank you for the quick response! Below is a snippet of my code, I am trying to pulse an atom with a Hadamard gate and then a Z rotation with phi equal to pi/2. However, my final state is,
 [[-0.70707771]
 [ 0.70713585]]

After removing the Z rotation, I get the same result (thus nothing is happening) on top of that, since my initial basis state is [[1],[0]], should I not be getting 
[[0.70707771]
 [ 0.70713585]]
Instead?

```
reg = Register({"q0": (0, 0)})
device = Chadoq2
seq = Sequence(reg, device)
seq.declare_channel("ch0", "raman_local", initial_target="q0")
    
def h_pulse(target,channel):
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi/2), detuning=0, phase=np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(np.pi, "q0", basis="digital")
    
def r_pulse(theta,target,channel):
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, 2*np.pi), detuning=0, phase=0)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(theta, "q0", basis="digital")

h_pulse("q0","ch0")
r_pulse(np.pi/2,"q0","ch0")

seq.draw(draw_phase_shifts=True)
sim = Simulation(seq,sampling_rate=0.1)
results = sim.run(progress_bar=True)
result_basis = results.get_final_state(reduce_to_basis = "digital")
print(result_basis)
```

**Reply by HGSilveri** [reply]
I'm afraid you won't be able to see the effect with a 2π-pulse, as that will always equate to applying an identity gate. The simplest way to see the effect of the phase shift is to try to do two Hadamard gates in a row and to enable/disable the π phase shift. You will see that, with the phase-shift, you will end up back in |0> (as expected), while without it you will end up in |1> (because you just did two π/2 rotations around -Y). 

The crucial point to consider here is that the virtual-Z gate approach is not concerned about the actual phase of the quantum state, only it's effect on the final outcome. In a way, their effect shows up only when it matters.

PS: If for some reason you want to prepare the |+> state, just do the rotation with `phase=-np.pi/2` (ie around the Y axis).

**Reply by juanboschero** [reply]
Oops I do not know why I thought having a 2π-pulse would work... However, let's say I apply a Y-pulse to my [[1],[0]] state, I then get a [[0],[1]]. I do this below and this works perfectly.

```
reg = Register({"q0": (0, 0)})
device = Chadoq2
seq = Sequence(reg, device)
seq.declare_channel("ch0", "raman_local", initial_target="q0")

def y_pulse(target,channel):
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi), detuning=0, phase= -np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)

y_pulse("q0","ch0")
seq.draw(draw_phase_shifts=True)
sim = Simulation(seq,sampling_rate=0.1)
results = sim.run(progress_bar=True)
result_basis = results.get_final_state(reduce_to_basis = "digital")
print(result_basis)
```
Result:
```
Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
Qobj data =
[[-8.76161965e-05]
 [ 9.99999996e-01]]
```
However, now if I want to apply a Z rotation to this state of π/2 to get a final state [[0],[i]], how would I go around doing that? I tried applying the double hadamard but it does not seem to change my final state,

```
reg = Register({"q0": (0, 0)})
device = Chadoq2
seq = Sequence(reg, device)
seq.declare_channel("ch0", "raman_local", initial_target="q0")

def y_pulse(target,channel):
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi), detuning=0, phase= -np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    
def r_pulse(theta,target,channel): 
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi/2), detuning=0, phase=np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(np.pi, "q0", basis="digital")
    
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi/2), detuning=0, phase=np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(np.pi, "q0", basis="digital")

    seq.phase_shift(theta, "q0", basis="digital")
    
y_pulse("q0","ch0")
r_pulse(np.pi/2,"q0","ch0")

seq.draw(draw_phase_shifts=True)
sim = Simulation(seq,sampling_rate=0.1)
results = sim.run(progress_bar=True)
result_basis = results.get_final_state(reduce_to_basis = "digital")
print(result_basis)
```
```
Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
Qobj data =
[[-8.77531281e-05]
 [ 9.99999996e-01]]
```

**Reply by HGSilveri** [reply]
`[0, i]` is actually the same as `[0, 1]`, they only differ by a global factor of `i`.

**Reply by juanboschero** [reply]
To apply a global factor would I need to introduce the global channel? Also, would applying a global ```i``` factor work for something like a controlled phase gate? I ask because I am trying to implement the phase estimation algorithm using pulses. Below is how I apply a CPHASE gate. I first pulse both qubits with a Y-pulse to have them in the |11> state, and then I make a controlled phase gate using the identity 

------------------0---------------------0
----Rz(theta)----+ -----Rz(-theta)----+

where,
 ---0--- is a CNOT gate in which 0 is the control and + is the target.
 ---+---

I implement the CNOT by using placing a Hadamard on the target followed by a controlled z between both qubits and then another Hadamard.  Below is the code,

```
q_dict = {"q0": (0, 0),"q1": (0, 4)}
reg = Register(q_dict)
device = Chadoq2
seq = Sequence(reg, device)
seq.declare_channel("ch0", "raman_local", initial_target="q0")
seq.declare_channel("ch1", "rydberg_local", initial_target="q0")

def ry_pulse(target,channel):
    pulse = define_pulse(250,np.pi,0,-np.pi/2,0)
    seq.target(target,channel)
    seq.add(pulse,channel)

def apply_cz(channel1,channel2,control,target):
    pi_wf = BlackmanWaveform(200,np.pi)
    pi_pulse = Pulse.ConstantDetuning(pi_wf,0,0)
    max_val = Chadoq2.rabi_from_blockade(8)
    two_pi_wf = BlackmanWaveform.from_max_val(max_val=max_val,area=2*np.pi)
    two_pi_pulse = Pulse.ConstantDetuning(amplitude=two_pi_wf,detuning=0,phase=0)
    seq.align(channel1,channel2)
    seq.target(control,channel2)
    seq.add(pi_pulse, channel2)
    seq.target(target,channel2)
    seq.add(two_pi_pulse, channel2)
    seq.target(control,channel2)
    seq.add(pi_pulse, channel2)
    seq.align(channel1,channel2)    

def apply_cnot(channel1,channel2,control,target):
    apply_hadamard(target,channel1)
    apply_cz(channel1,channel2,control,target)
    apply_hadamard(target,channel1)

def apply_zrot(theta,target,channel): 
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi/2), detuning=0, phase=np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(np.pi, "q0", basis="digital")
    
    pulse = Pulse.ConstantDetuning(BlackmanWaveform(250, np.pi/2), detuning=0, phase=np.pi / 2)
    seq.target(target,channel)
    seq.add(pulse,channel)
    seq.phase_shift(np.pi, "q0", basis="digital")

    seq.phase_shift(theta, "q0", basis="digital")

def apply_phase(theta,channel1,channel2,control,target):
    apply_zrot(theta,target,channel1)
    apply_cnot(channel1,channel2,control,target)
    apply_zrot(-theta,target,channel1)
    apply_cnot(channel1,channel2,control,target)

ry_pulse("q0","ch0")
ry_pulse("q1","ch0")
apply_phase(np.pi/4,"ch0","ch1","q0","q1")

seq.draw(draw_phase_shifts=True)

sim = Simulation(seq,sampling_rate=0.5)
results = sim.run(progress_bar=True)
result_basis = results.get_final_state(reduce_to_basis = "digital", tol = 1)
print(result_basis)
```
In this situation I get [[0],[0],[0],[1]] but if I apply a global phase, I would essentially add the phase to all states regardless if my basis was in the |11> state, right?

Thank you again for your time.

**Reply by HGSilveri** [reply, marked as answer]
When we speak of two states (|phi_1>, |phi_2>) being equal up to a global phase, it just means there is some phase `alpha` that makes 
`|phi_1> = exp(i*alpha) |phi_2>`. In QC this global phase factors are always ignored because, when measuring some observable `Ô`, `<phi_1| Ô |phi_1> = <phi_1| exp(-i * alpha) Ô exp(i*alpha) |phi_1> = <phi_2| Ô |phi_2>`, so it can always be ignored.

What you are asking for (adding a global phase to a state) is something that is never considered in QC and there is no gate that will do so.

I understand that you are trying to see the effect of a CPHASE gate, but it will never show in an eigenstate. You need to have your state be in a superposition of basis states to see an effect. For example, if you start with

`(|00> + |01> + |10> + |11>)/2`

and apply `CPHASE(theta)` you will get

`(|00> + |01> + exp(i*theta)|10> + exp(i*theta)|11>)/2`

Note how the phase acquired is not global (ie it can't be factored out as a multiplicative factor on all states) and so it can't be ignored anymore. 

Nonetheless, if you try this in Pulser with virtual-Z gates you might still not see this phase appear in the state vector for the same reason as before - phase gates before a measurement don't affect the result and are thus not applied. You need to apply another gate afterwards that relies on the fact that the CPHASE gate worked to reach some result (similary to what I suggested you tried out with the two Hadamards).

**Comment by juanboschero** [comment]
Hi @HGSilveri,
I think the issue I had was not fully understanding the fundamentals of tensor measurements.  Moreover, I realized that the way I defined my CPHASE was wrong, by applying [this](https://physics.stackexchange.com/questions/213002/in-quantum-fourier-transform-why-can-any-controlled-r-k-gate-be-formed-by-t/213168#213168) instead, I managed to properly make the complete algorithm. Thanks again for all the help!
